src/util/compat.test.js

```diff
@@ -0,0 +1,169 @@
+// @flow
+
+import {toCompat, fromCompat} from "./compat";
+import type {Compatible} from "./compat";
+
+describe("compat utilities", () => {
+  type FooV1 = {foo: number};
+  type FooV2 = {bar: {foo: number}};
+  const type = "Foo Plugin's Experiment";
+  const v1 = "v1";
+  const v2 = "v2";
+  function fooHandler(x: FooV1): FooV2 {
+    return {bar: x};
+  }
+  const example = () => {
+    const dataV1: FooV1 = {foo: 1};
+    const dataV2: FooV2 = {bar: {foo: 1}};
+    const compatV1 = toCompat(type, v1, dataV1);
+    const compatV2 = toCompat(type, v2, dataV2);
+    return {dataV1, dataV2, compatV1, compatV2};
+  };
+  it("toCompat adds a compat field", () => {
+    const {dataV1, compatV1} = example();
+    expect(dataV1).toEqual({foo: 1}); // calling compat caused no mutation
+    const {type: actualType, version: actualVersion} = (compatV1: any)[0];
+    expect(actualType).toEqual(type);
+    expect(actualVersion).toEqual(v1);
+  });
+
+  it("toCompat doesn't fails on primitives", () => {
+    expect(fromCompat(type, v1, toCompat(type, v1, 3))).toBe(3);
+  });
+
+  it("toCompat -> fromCompat is identity", () => {
+    const {dataV1, compatV1} = example();
+    expect(fromCompat(type, v1, compatV1)).toEqual(dataV1);
+  });
+
+  it("fromCompat fails if compatibility undefined", () => {
+    const dataV1: any = example().dataV1;
+    expect(() => fromCompat(type, v1, dataV1)).toThrowError(
+      "didn't have compatibility defined"
+    );
+  });
+
+  it("fromCompat fails if type is wrong", () => {
+    const {compatV1} = example();
+    expect(() => fromCompat("who is Foo?", v1, compatV1)).toThrowError(
+      "Expected type"
+    );
+  });
+
+  it("fromCompat fails if version is wrong", () => {
+    const {compatV1} = example();
+    expect(() => fromCompat(type, v2, compatV1)).toThrowError(
+      "unsupported version"
+    );
+  });
+
+  it("handlers can load older versions", () => {
+    const {compatV1, dataV2} = example();
+    const handlers = {v1: fooHandler};
+    expect(fromCompat(type, v2, compatV1, handlers)).toEqual(dataV2);
+  });
+
+  it("handlers activate even on current version", () => {
+    const {compatV1} = example();
+    const handlers = {
+      v1: () => ({
+        hello: "world",
+      }),
+    };
+    expect(fromCompat(type, v1, compatV1, handlers)).toEqual({hello: "world"});
+  });
+
+  describe("composable versioning", () => {
+    class InnerV1 {
+      x: number;
+      constructor(x: number) {
+        this.x = x;
+      }
+      toJSON(): Compatible<FooV1> {
+        return toCompat("inner", v1, {foo: this.x});
+      }
+      static fromJSON(json): InnerV1 {
+        const from: FooV1 = fromCompat("inner", v1, json);
+        return new InnerV1(from.foo);
+      }
+    }
+
+    class InnerV2 {
+      x: number;
+      constructor(x: number) {
+        this.x = x;
+      }
+      toJSON(): Compatible<FooV2> {
+        return toCompat("inner", v2, {bar: {foo: this.x}});
+      }
+
```

Comment:
nit: blank line inconsistency between `InnerV1` and `InnerV2`

Reply:
Fixed